Expand glob import on cyclic import fail

<!--
Troubleshooting guide: https://rust-analyzer.github.io/book/troubleshooting.html
Forum for questions: https://users.rust-lang.org/c/ide/14

Before submitting, please make sure that you're not running into one of these known issues:

 1. on-the-fly diagnostics are mostly unimplemented (`cargo check` diagnostics will be shown when saving a file): #3107

Otherwise please try to provide information which will help us to fix the issue faster. Minimal reproducible examples with few dependencies are especially lovely <3.
-->

**rust-analyzer version**: rust-analyzer 1.86.0 (05f9846 2025-03-31)

**rustc version**: rustc 1.86.0 (05f9846f8 2025-03-31)

**editor or extension**: VIM9.1 coc-rust-analyzer

**code snippet to reproduce**:
```rust
mod foo {
    pub use crate::*;
    pub struct Foo;
    pub fn bar() -> Bar { _ = Foo; Bar }
}
pub use foo::*;
pub struct Bar;
```

On line 2 use `Expand glob import`:

```rust
mod foo {
    pub use crate::{Bar, Foo, foo};
    pub struct Foo;
    pub fn bar() -> Bar { _ = Foo; Bar }
}
pub use foo::*;
pub struct Bar;
```

`cargo check` outputs:

```rust
    Checking test2 v0.1.0 (/home/lrne/Project/Rust/test2)
error[E0255]: the name `Foo` is defined multiple times
 --> src/main.rs:3:5
  |
2 |     pub use crate::{Bar, Foo, foo};
  |                          --- previous import of the type `Foo` here
3 |     pub struct Foo;
  |     ^^^^^^^^^^^^^^^ `Foo` redefined here
  |
  = note: `Foo` must be defined only once in the type namespace of this module
help: you can use `as` to change the binding name of the import
  |
2 |     pub use crate::{Bar, Foo as OtherFoo, foo};
  |                              +++++++++++

error[E0365]: `foo` is only public within the crate, and cannot be re-exported outside
 --> src/main.rs:2:31
  |
2 |     pub use crate::{Bar, Foo, foo};
  |                               ^^^ re-export of crate public `foo`
  |
  = note: consider declaring type or module `foo` with `pub`

warning: unused imports: `Foo` and `foo`
 --> src/main.rs:2:26
  |
2 |     pub use crate::{Bar, Foo, foo};
  |                          ^^^  ^^^
  |
  = note: `#[warn(unused_imports)]` on by default

Some errors have detailed explanations: E0255, E0365.
For more information about an error, try `rustc --explain E0255`.
warning: `test2` (bin "test2") generated 1 warning
error: could not compile `test2` (bin "test2") due to 2 previous errors; 1 warning emitted
```

**Expect**:

```rust
mod foo {
    pub use crate::Bar;
    pub struct Foo;
    pub fn bar() -> Bar { _ = Foo; Bar }
}
pub use foo::*;
pub struct Bar;
```

`Foo` and `foo` defined this module

--- a/crates/ide-assists/src/handlers/expand_glob_import.rs
+++ b/crates/ide-assists/src/handlers/expand_glob_import.rs
@@ -162,7 +162,8 @@ fn build_expanded_import(
         }
     };
 
-    let refs_in_target = find_refs_in_mod(ctx, target_module, visible_from, must_be_pub);
+    let refs_in_target =
+        find_refs_in_mod(ctx, target_module, current_module, visible_from, must_be_pub);
     let imported_defs = find_imported_defs(ctx, use_item);
 
     let filtered_defs =
@@ -303,6 +304,7 @@ impl Refs {
 fn find_refs_in_mod(
     ctx: &AssistContext<'_, '_>,
     expandable: Expandable,
+    current_module: Module,
     visible_from: Module,
     must_be_pub: bool,
 ) -> Refs {
@@ -312,6 +314,10 @@ fn find_refs_in_mod(
             let refs = module_scope
                 .into_iter()
                 .filter_map(|(n, d)| Ref::from_scope_def(ctx, n, d))
+                .filter(|r| match r.def {
+                    Definition::Module(it) => it != current_module,
+                    _ => r.def.module(ctx.db()).map_or(false, |it| it != current_module),
+                })
                 .filter(|r| !must_be_pub || r.is_pub)
                 .collect();
             Refs(refs)
@@ -439,6 +445,31 @@ use foo::{Bar, Baz, f};
 fn qux(bar: Bar, baz: Baz) {
     f();
 }
+",
+        )
+    }
+
+    #[test]
+    fn expanding_glob_import_on_cycle_import() {
+        check_assist(
+            expand_glob_import,
+            r"
+mod foo {
+    pub use crate::*$0;
+    pub struct Foo;
+    pub fn bar() -> Bar { _ = Foo; Bar }
+}
+pub use foo::*;
+pub struct Bar;
+",
+            r"
+mod foo {
+    pub use crate::Bar;
+    pub struct Foo;
+    pub fn bar() -> Bar { _ = Foo; Bar }
+}
+pub use foo::*;
+pub struct Bar;
 ",
         )
     }

